google/cloud/spanner/internal/partial_result_set_source.cc

```diff
@@ -14,186 +14,243 @@
-    chunk_ = {};
+    ExtractSubrangeAndAppend(new_values, append_start, values_);
+    values_back_incomplete_ = result_set->result.chunked_value();
   }
 
-  if (result_set->result.chunked_value()) {
-    if (new_values.empty()) {
-      return Status(StatusCode::kInternal,
-                    "PartialResultSet had chunked_value "
-                    "set true but contained no values");
+  // Deliver whatever rows we can muster.
+  auto const n_values = values_.size() - (values_back_incomplete_ ? 1 : 0);
+  auto const n_columns = columns_ ? static_cast<int>(columns_->size()) : 0;
+  auto n_rows = n_columns ? n_values / n_columns : 0;
+  if (n_columns == 0 && !values_.empty()) {
+    return Status(StatusCode::kInternal,
+                  "PartialResultSetSource metadata is missing row type");
+  }
+
+  // If we didn't receive a resume token, and have not exceeded our buffer
+  // limit, then we choose to `Read()` again so as to maintain resumability.
+  if (result_set->result.resume_token().empty()) {
+    if (values_.SpaceUsedExcludingSelfLong() < values_space_limit_) {
+      return {};  // OK
     }
-    chunk_ = std::move(new_values[new_values.size() - 1]);
-    new_values.RemoveLast();
   }
 
-  // Moves all the remaining in new_values to buffer_
-  for (auto& value_proto : new_values) {
-    buffer_.push_back(std::move(value_proto));
+  // If we did receive a resume token then everything should be deliverable,
+  // and we'll be able to resume the stream at this point after a breakage.
+  // Otherwise, if we deliver anything at all, we must disable resumability.
+  if (!result_set->result.resume_token().empty()) {
+    resume_token_ = result_set->result.resume_token();
+    if (n_rows * n_columns != values_.size()) {
+      if (state_ != kEndOfStream) {
+        return Status(StatusCode::kInternal,
+                      "PartialResultSetSource reader produced a resume token"
+                      " that is not on a row boundary");
+      }
+      if (n_rows == 0) {
+        return Status(StatusCode::kInternal,
+                      "PartialResultSetSource stream ended at a point"
+                      " that is not on a row boundary");
+      }
+    }
+  } else if (n_rows != 0) {
+    resume_token_ = absl::nullopt;
+  }
+
+  // Combine the available values into new elements of `rows_`.
+  int values_pos = 0;
+  std::vector<spanner::Value> values;
+  values.reserve(n_columns);
+  for (; n_rows != 0; --n_rows) {
+    for (auto const& field : metadata_->row_type().fields()) {
+      auto& value = *values_.Mutable(values_pos++);
+      values.push_back(FromProto(field.type(), std::move(value)));
+    }
+    rows_.push_back(RowFriend::MakeRow(std::move(values), columns_));
+    values.clear();
   }
 
+  // If we didn't combine all the values, leave the remainder for next time.
+  auto* rem_values = result_set->result.mutable_values();
+  ExtractSubrangeAndAppend(values_, values_pos, *rem_values);
+  values_.Swap(rem_values);
+
```

Comment:
Here I am confused. Don't we already move out all the values and clear the source, so `result_set->result.values().empty() == true` ? Can we remove this code?
https://github.com/googleapis/google-cloud-cpp/blob/441dba6359ab5a2997cfc2c4bff487414ae2648b/google/cloud/spanner/internal/partial_result_set_source.cc#L187
https://github.com/googleapis/google-cloud-cpp/blob/441dba6359ab5a2997cfc2c4bff487414ae2648b/google/cloud/spanner/internal/partial_result_set_source.cc#L194
https://github.com/googleapis/google-cloud-cpp/blob/441dba6359ab5a2997cfc2c4bff487414ae2648b/google/cloud/spanner/internal/partial_result_set_source.cc#L53

Reply:
Ah, thanks for explaining. I see it now.

Reply:
Just as a further clarification, if we get to here then `values_pos` is probably equal to `values_.size()`.  That is, we've combined all the values into rows, and there is nothing left over.  The exception is when we've exceeded the buffer size, and yielded rows without having a resume token to suppress their replay.

Reply:
And P.S. ... thanks for slogging through all of that!